src/python/pants/backend/python/subsystems/python_tool_base.py

```diff
@@ -47,6 +54,43 @@ class PythonToolRequirementsBase(Subsystem):
     default_lockfile_resource: ClassVar[tuple[str, str] | None] = None
     default_lockfile_url: ClassVar[str | None] = None
 
+    install_from_resolve = StrOption(
+        advanced=True,
+        default=None,
+        help=lambda cls: softwrap(
+            f"""\
+            If specified, install the tool using the lockfile for this named resolve.
+
+            This resolve must be defined in [python].resolves, as described in
+            {doc_url("python-third-party-dependencies#user-lockfiles")}, and its lockfile must
+            provide the requirements named in the `requirements` option.
+
+            If unspecified, and the `lockfile` option is unset, the tool will be installed
+            using the default lockfile shipped with Pants.
+
+            If unspecified, and the `lockfile` option is set, the tool will use the custom
+            `{cls.options_scope}` "tool lockfile" generated from the `version` and
+            `extra_requirements` options. But note that this mechanism will soon be deprecated.
+            """
+        ),
+    )
+    # TODO: After we deprecate and remove the tool lockfile concept, we can remove the
+    #  version and extra_requirements options and directly list loosely-constrained
+    #  requirements for each tool in this option's default. The specific versions will then
+    #  come either from the default lockfile we provide, or from a user lockfile.
+    requirements = StrListOption(
+        advanced=True,
+        default=lambda cls: cls.default_requirements
+        or cls.unversioned_requirements(
```

Comment:
This is bad right? Stripping the version specifier allows us to pick up a version that does not support the CLI args we use or have the semantics we've constrained via version specifiers.

Reply:
Tool requirements can come from a user resolve, but they have to still be compatible with what our rule code expects or else disaster at a distance.

Reply:
Technically yes, but that is also true prior to this change - I can set the version (and extra_requirements) to whatever I want and generate a (now old-style) "tool lockfile" off that.

Right now our tool requirements tend to be overly constrained, e.g., [black](https://github.com/pantsbuild/pants/blob/5f3f251a378f780daac07e0e5409c9a132571a89/src/python/pants/backend/python/lint/black/subsystem.py#L54), [pytest](https://github.com/pantsbuild/pants/blob/d39ad4b252578e20bc0f9e154607be165afb4949/src/python/pants/backend/python/subsystems/pytest.py#L78), or, slightly more permissively, but not much, [isort](https://github.com/pantsbuild/pants/blob/7251de9bb8aef692b4ffcfc62c36943983a51145/src/python/pants/backend/python/lint/isort/subsystem.py#L31). 

So if we use these as-is, users will have to tinker with `version` (and/or `extra_requirements`) to loosen them up as well as specify versions (or possibly tighter version ranges) for them in some requirements.txt.

We could audit all the tools and specify loose requirements directly on `default_requirements`, but perhaps in a followup? There are a lot of them. 

And this also raises the question of how would we do so? We can pick a minimum version that we know we're compatible with, but how would we pick an upper-bound? In some cases we can assume semver, perhaps, so we can bound at the next major version. But then that may be quite restrictive, because in practice it might be likely that Pants will work with the next major version, and we've forced users, again, to update `requirements` as well as a requirements.txt. And the whole point here was to avoid that sort of near-repetition. Granted they might not need to update `requirements` as often as the more specific version in requirements.txt, perhaps only when they want to bump the major version.

So I'm a little torn about what the right thing to do is.

Reply:
I like this:
https://github.com/pantsbuild/pants/blob/d68cc258e775db1522578a53d6b63870b9b45d34/src/python/pants/backend/python/util_rules/pex_cli.py#L41-L43

Now that's for downloaded tools and not resolved tools of course, but it gets a few things right:
+ a lower bound is set: we know we need the features in that lower bound, you can't go lower, we'd break
+ it allows upgrades: the context is Pex and so we know Pex doesn't break people so we can bound at a semver major bump. For any given tool we should have some sense of this - are they cowboys or reasonable or what have you.

I think it would be good to get to this same place with tool resolves. If we guessed to low an upper bound the user has to double adjust, but there's an out. We should get good at guessing an upper bound though, aka we should actually know the tools we support well.

Reply:
Maybe, as a decent heuristic, we can lower-bound to something reasonable that we know we support, and not upper bound, under the assumption that 90% of the time options we use or assumptions we make will still hold in later tool versions, and in the case where they don't, well, then there will be an error and the user will have to infer from the error message that Pants isn't compatible with their selected tool version.

Taking to logical extremes, we can make almost no guarantees about any version of any tool other than the ones we test on. So we'll have to pick some middle ground, and maybe that is an OK one.

Reply:
I wrote this before I saw your comment right above it. Yeah, so this will require an audit of all the tools. I'll take a look at that in the morning.

Reply:
Maybe. I still think we should actually know the tools we support. When you do, I think it's not difficult to pick an upper bound (or not). It's only when you don't actually know your tools well that you have to do something like punt and pick no upper bound.

Maybe punting is fine since we don't know our tools well as a rule. Paired with high level doc about buyer beware, maybe this won't become a support foot gun.

Reply:
OK, switched this to using the restricted requirements, and we can loosen them up on a case-by-case basis in a followup.